Sources/Commands/PackageTools/PluginCommand.swift

```diff
@@ -116,38 +126,64 @@ struct PluginCommand: SwiftCommand {
-                        let answer = readLine(strippingNewline: true)
-                        // Throw an error if we didn't get permission.
-                        if answer?.lowercased() != "yes" {
-                            throw ValidationError("Plugin was denied permission to write to the package directory.")
-                        }
-                        // Otherwise append the directory to the list of allowed ones.
-                        writableDirectories.append(package.path)
-                    }
-                    else {
-                        // We can't ask the user, so emit an error suggesting passing the flag.
-                        let remedy = "Use `--allow-writing-to-package-directory` to allow this."
-                        throw ValidationError([problem, reason, remedy].joined(separator: "\n"))
+
+        // If the plugin requires permissions, we ask the user for approval.
+        if case .command(_, let permissions) = plugin.capability {
+            try permissions.forEach {
+                let permissionString: String
+                let reasonString: String
+                let remedyOption: String
+
+                switch $0 {
+                case .writeToPackageDirectory(let reason):
+                    guard !options.allowWritingToPackageDirectory else { return } // permission already granted
+                    permissionString = "write to the package directory"
+                    reasonString = reason
+                    remedyOption = "--allow-writing-to-package-directory"
+                case .allowNetworkConnections(let scope, let reason):
+                    guard scope != .none else { return } // no need to prompt
+                    guard options.allowNetworkConnections != scope else { return } // permission already granted
+                    let portsString = scope.ports.isEmpty ? "on all ports" : "on ports: \(scope.ports)"
+                    permissionString = "allow \(scope.label) network connections \(portsString)"
+                    reasonString = reason
+                    remedyOption = "--allow-network-connections \(scope.label)"
+                }
+
+                let problem = "Plugin ‘\(plugin.name)’ wants permission to \(permissionString)."
+                let reason = "Stated reason: “\(reasonString)”."
+                if swiftTool.outputStream.isTTY {
+                    // We can ask the user directly, so we do so.
+                    let query = "Allow this plugin to \(permissionString)?"
+                    swiftTool.outputStream.write("\(problem)\n\(reason)\n\(query) (yes/no) ".utf8)
+                    swiftTool.outputStream.flush()
+                    let answer = readLine(strippingNewline: true)
+                    // Throw an error if we didn't get permission.
+                    if answer?.lowercased() != "yes" {
+                        throw StringError("Plugin was denied permission to \(permissionString).")
```

Comment:
Note: I changed these to `StringError` to avoid having usage be printed to the console which seems to be a behavior of ArgumentParser's `ValidationError`.